packages/eslint-plugin/src/rules/shorthand-property-sorting/index.ts

```diff
@@ -0,0 +1,140 @@
+  context.report({
+    node: node,
+    messageId: 'shorthand-first',
+    fix: (fixer) => {
+      if (node.type === 'ObjectExpression' && node.properties.length > 0) {
+        // sort the properties by depth
+        const sortedProperties = node.properties.slice().sort((a, b) => {
+          if (
+            a.type === 'Property' &&
+            a.key.type === 'Identifier' &&
+            b.type === 'Property' &&
+            b.key.type === 'Identifier'
+          ) {
+            const propA = kebabCase(a.key.name) as ShorthandProperties;
+            const propB = kebabCase(b.key.name) as ShorthandProperties;
+
+            return shorthandBuckets[propA] - shorthandBuckets[propB];
+          }
+          return 0;
+        });
+
+        const sourceCode = context.getSourceCode();
+        const sortedCode = sortedProperties
+          .map((property) => sourceCode.getText(property))
+          .join(', ');
+
+        // Replace the old object expression with the new sorted one
+        const newObjectExpression = `{ ${sortedCode} }`;
+
+        return fixer.replaceText(node, newObjectExpression);
+      }
+      return null;
+    },
+  });
+};
+
+const callExpressionCorrectImport = (node: Rule.Node, references: Scope.Reference[]): boolean => {
+  return (
+    node.type === 'CallExpression' &&
+    (isCss(node.callee as Rule.Node, references) ||
+      isCssMap(node.callee as Rule.Node, references) ||
+      isStyled(node.callee as Rule.Node, references))
+  );
+};
+
+export const shorthandFirst: Rule.RuleModule = {
+  meta: {
+    docs: {
+      recommended: true,
+      description:
+        "At build time, Compiled automatically sorts shorthand properties (like `font` and `border`) so that they come before any longhand properties (like `fontSize` and `borderTopColor`) defined on the component. This means that longhand properties will always override shorthand properties. This rule enforces that the order in which the properties appear in a component's source code matches the actual ordering the properties will have at build time and runtime.",
+      url: 'https://github.com/atlassian-labs/compiled/tree/master/packages/eslint-plugin/src/rules/shorthand-property-sorting',
+    },
+    messages: {
+      'shorthand-first':
+        'When using both shorthand and longhand properties, the shorthand property should be first.',
+    },
+    type: 'suggestion',
+    fixable: 'code',
+    hasSuggestions: true,
+  },
+  create(context) {
+    const selectorString: string =
+      'CallExpression[callee.type="Identifier"][callee.name="css"] > ObjectExpression,' +
+      'CallExpression[callee.type="Identifier"][callee.name="cssMap"] > ObjectExpression,' +
+      'CallExpression[callee.type="Identifier"][callee.name="xcss"] > ObjectExpression,' +
+      'CallExpression[callee.type="MemberExpression"][callee.object.name="styled"] > ObjectExpression';
+
+    return {
+      [selectorString]: (node: Rule.Node) => {
```

Comment:
Great job! This enforcement includes several cases to consider, such as:

Styles Composition
```
const foo = css({ margin: ... });
const bar = css({ marginTop: ... });
<div css={[bar, foo]} />;
```
Style Inheritance
```
const Foo = styled.div({ marginTop: ... });
const Bar = styled(Foo)({ margin: ... });
```
I don't think we need to address all these cases in this pull request, I'm mentioning them here so we can improve the linting rule in future iterations.